### PR TITLE
Change usage semantics

### LINKED ISSUE #970: Change Usage semantics
The current `Usage` enum is heavily inspired by DX11. There is nothing wrong with it in particular, it maps nicely to both DX11 and GL. But looking forward, we need more flexibility, since other APIs are not locked to these usage types. I propose a set of flags instead:

``` rust
bitflags!(
    /// CPU transfer operations.
    /// In general, CPU access (read/write) forces a texture to have linear tiling, which reduces the GPU access pattern efficiency (hence, the rendering performance). Addition costs apply to synchronize CPU-GPU interaction by the driver.
    /// It is recommended though (if not enforced in the future) to not have renderable textures that are read/write. Instead, you create 2 separate textures - one for GPU only rendering with `TRANS_COPY` flag. Another - for CPU-only read/write texture ("staging texture" in DX11) that you will copy.
    pub flags Transfer: u8 {
        /// Resource is readable by CPU.
        pub TRANS_READ = 0x1,
        /// Resource is writable by CPU.
        pub TRANS_WRITE = 0x2,
        /// Resource can be copied from/into CPU visible memory. This flag barely affects performance.
        pub TRANS_COPY = 0x4,
    }
);
```

The backend will be able to derive the appropriate internal usage type from the `(Transfer, Bind)` flags. If a resource doesn't have any flags, it's considered immutable (in DX11 terms). `TRANS_WRITE` alone means the resource is dynamic. Both read/write mean the resource is staging. The backends are free to error out on an incompatible set of flags.

### PR DESCRIPTION
**Pull request by Bastacyclop**
Planned:
- [x] Modify core and OpenGL backend
- [x] Simplify the mapping code
- [x] Modify the other backends
- [x] Update gfx's examples and run them

Fixes #970
#423 will become necessary for staging.

### CODE REVIEW

File: src/render/src/factory.rs [hunk 1]
```diff
@@ -125,51 +125,40 @@ pub trait FactoryExt<R: Resources>: Factory<R> {
...
+    fn create_constant_buffer<T>(&mut self, num: usize)
+                                 -> (handle::Buffer<R, T>,
+                                     mapping::WritableOnly<R, T>)
         where T: Copy
     {
-        let buffer = self.create_buffer_mappable(num, role, bind, memory::READ)
-            .unwrap();
-        let mapping = self.map_buffer_readable(&buffer).unwrap();
-        (buffer, mapping)
+        self.create_upload_buffer(num, buffer::Role::Constant, Bind::empty())
```
Comment (Bastacyclop) [inline src/render/src/factory.rs]: Or create a `Data` buffer with `TRANSFER_DST` ?

File: src/core/src/buffer.rs [hunk 2]
```diff
@@ -103,6 +103,8 @@ pub enum Role {
     Index,
     /// Constant buffer
     Constant,
+    /// Staging buffer
+    Staging,
```
Comment (kvark) [inline src/core/src/buffer.rs]: I don't think it makes sense to be in `Role`. Staging is a property of underlying memory, not the role. Any role buffer can be stages.

File: src/backend/gl/src/factory.rs [hunk 3]
```diff
@@ -143,19 +147,9 @@ impl Factory {
...
-                    memory::READ => gl::DYNAMIC_READ,
-                    memory::WRITE => gl::DYNAMIC_DRAW,
-                    _ => unreachable!(),
-                },
-                CpuOnly(access) => match access {
-                    memory::READ => gl::STREAM_READ,
-                    _ => gl::DYNAMIC_DRAW,
-                }
+                Data => gl::STATIC_DRAW,
+                Upload => gl::DYNAMIC_DRAW,
```
Comment (kvark) [inline src/backend/gl/src/factory.rs]: not `STREAM_DRAW`?

File: src/core/src/mapping.rs [hunk 4]
```diff
@@ -165,6 +166,8 @@ impl<'a, R: Resources, T: 'a + Copy> DerefMut for RWer<'a, R, T> {
     fn deref_mut(&mut self) -> &mut [T] { self.slice }
 }
 
+// TODO: should we keep the RW mapping stuff ?
```
Comment (kvark) [inline src/core/src/mapping.rs]: I say kill it ;)

File: src/render/src/factory.rs (hunk 1 above)
Comment (kvark) [inline src/render/src/factory.rs]: Yes, I think the constant buffers should be `Data`. Their CPU copies are attached to the command buffers, after all

File: src/core/src/buffer.rs (hunk 2 above)
Comment (Bastacyclop) [inline src/core/src/buffer.rs]: What happens when a buffer is not used for anything but staging ? It does not make much sense for it to be any of the current roles.

Comment (Bastacyclop) [inline src/core/src/buffer.rs]: Maybe a `None` variant ?

Comment (kvark) [inline src/core/src/buffer.rs]: I think you are right. Staging buffers don't fulfil any role directly (other than being staging, obviously), so your initial approach is correct. For GL, I suspect the staging buffer will really be a PBO target.

File: src/backend/gl/src/factory.rs [hunk 5]
```diff
@@ -148,7 +148,7 @@ impl Factory {
         else {
             let usage = match info.usage {
                 Data => gl::STATIC_DRAW,
```
Comment (kvark) [inline src/backend/gl/src/factory.rs]: Warning: that would be inefficient for our constant buffers, which change often

Comment (Bastacyclop) [inline src/backend/gl/src/factory.rs]: I don't see any way to know if the data is going to change often with the new usage, should we reintroduce the `Dynamic` variant then ?

Comment (Bastacyclop) [inline src/backend/gl/src/factory.rs]: Another way could be to use `Upload` + mapping hidden behind the update commands.

Comment (Bastacyclop) [inline src/backend/gl/src/factory.rs]: I think a `Dynamic` variant is more flexible for the backend side.

Comment (kvark) [inline src/backend/gl/src/factory.rs]: I thought about it some more, let it be `STATIC_DRAW`, see the [discussion](https://gitter.im/gfx-rs/gfx?at=58724f1cc02c1a3959fedc57) in gitter.

File: src/backend/dx11/src/factory.rs [hunk 6]
```diff
@@ -962,32 +930,29 @@ pub fn ensure_mapped(pointer: &mut *mut ::std::os::raw::c_void,
             DepthPitch: 0,
         };
             
-            
+        let dst = raw_handle.to_resource() as *mut winapi::d3d11::ID3D11Resource;
+        let map_type = winapi::d3d11::D3D11_MAP_WRITE; // no DISCARD because we are STAGING
```
Comment (Bastacyclop) [inline src/backend/dx11/src/factory.rs]: I will need to put `MAP_READ` here when mapping for read.

File: src/backend/dx11/src/execute.rs [hunk 7]
```diff
@@ -14,48 +14,40 @@
...
-                let dst = (sub.pData as *mut u8).offset(offset_bytes as isize);
-                ptr::copy_nonoverlapping(data.as_ptr(), dst, data.len());
-                (*context).Unmap(dst_resource, 0);
-                hr
-            };
-            if !winapi::SUCCEEDED(hr) {
-                error!("Buffer {:?} failed to map, error {:x}", buffer, hr);
-            }
-        },
+    /* TODO: do we want to allow updates for Data ?
```
Comment (Bastacyclop) [inline src/backend/dx11/src/execute.rs]: I will remove this

File: src/backend/dx11/src/execute.rs [hunk 8]
```diff
@@ -77,32 +69,40 @@ pub fn update_texture(context: *mut winapi::ID3D11DeviceContext, texture: &Textu
...
-                back:   (image.zoffset + image.depth) as UINT,
-            };
-            let ptr = data.as_ptr() as *const _;
-            unsafe {
-                //let subres = winapi::D3D11CalcSubresource(image.mipmap, array_slice, num_mipmap_levels);
-                (*context).UpdateSubresource(dst_resource, subres, &dst_box, ptr, row_pitch, depth_pitch)
-            };
-        },
-        Usage::Dynamic | Usage::CpuOnly(_) | Usage::Mappable(_) => unimplemented!(),
+    /* TODO: do we want to allow updates for Data ?
```
Comment (Bastacyclop) [inline src/backend/dx11/src/execute.rs]: I will remove this too

File: src/backend/dx11/src/execute.rs [hunk 9]
```diff
@@ -77,32 +69,40 @@ pub fn update_texture(context: *mut winapi::ID3D11DeviceContext, texture: &Textu
...
+        back:   (image.zoffset + image.depth) as UINT,
+    };
+    let ptr = data.as_ptr() as *const _;
+    unsafe {
+        //let subres = winapi::D3D11CalcSubresource(image.mipmap, array_slice, num_mipmap_levels);
+        (*context).UpdateSubresource(dst_resource, subres, &dst_box, ptr, row_pitch, depth_pitch)
+    };
+    */
+
+    // FIXME: probably wrong
```
Comment (Bastacyclop) [inline src/backend/dx11/src/execute.rs]: I need to verify this part, I think it did not take many details into account.

**Comment by Bastacyclop** [comment]
`create_buffer_immutable` may not produce an immutable buffer if the bind flags are "mutable".

**Comment by Bastacyclop** [comment]
@kvark What do you think about it ? I think it's ready for merge (publishing should probably wait a bit), I changed dx11's [texture update](https://github.com/gfx-rs/gfx/pull/1123/files#diff-e53e2203bb8a16e2d7c3c80d560bb067R58) without actually testing it. `Upload` and `Download` are basically unusable and untestable without #423.

**Comment by kvark** [comment]
@Bastacyclop thanks! I'll do another review pass today, and then we need to make sure it still works on Windows.
Then we'll probably jump on #423 before publishing.

**Comment by Bastacyclop** [comment]
I ran the examples on linux and windows and did not see any issue (except that "instancing" cannot run), but there is no texture update in them (I think).

File: examples/instancing/main.rs [hunk 10]
```diff
@@ -112,15 +112,24 @@ impl<R: gfx::Resources> gfx_app::Application<R> for App<R> {
...
 
-        let (instance_buffer, mut instance_mapping) = factory
-            .create_mapped_buffer_writable(MAX_INSTANCE_COUNT,
-                                     gfx::buffer::Role::Vertex,
-                                     gfx::Bind::empty());
+        let staging = factory
+            .create_buffer(MAX_INSTANCE_COUNT,
+                           gfx::buffer::Role::Staging,
+                           gfx::memory::Usage::Upload,
+                           gfx::TRANSFER_SRC).unwrap();
```
Comment (kvark) [inline examples/instancing/main.rs]: we might need a helper for the upload buffer constructor

**Comment by kvark** [comment]
@homu r+

**Pull request merged**